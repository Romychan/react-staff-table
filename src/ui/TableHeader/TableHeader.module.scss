.header {
  @apply bg-gray-50 dark:bg-gray-700;
}

.row {
}
