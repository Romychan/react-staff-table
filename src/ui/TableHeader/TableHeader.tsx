import { useState } from 'react';

import {
  TableOrderValue,
  IColumnTable,
  ISortingTable,
} from '@interfaces/index';

import { TableHeaderCell } from '../TableHeaderCell/TableHeaderCell';

import styles from './TableHeader.module.scss';

interface ITableHeaderProps<T, K extends keyof T> {
  /** Sets table columns */
  columns: IColumnTable<T, K>[];
  /**  Sets parameters for sorting */
  sorting: ISortingTable<T, K>;
  /** The callback that will be called when sorting is selected. */
  handleSorting: (accessor: K, order: TableOrderValue) => void;
}

export const TableHeader = <T, K extends keyof T>({
  columns,
  sorting,
  handleSorting,
}: ITableHeaderProps<T, K>) => {
  const [sortKey, setSortKey] = useState<K>(sorting.column);
  const [order, setOrder] = useState(sorting.order);

  const handleSortingChange = (accessor: K) => {
    const sortOrder = order === 'asc' ? 'desc' : 'asc';

    setSortKey(accessor);
    setOrder(sortOrder);
    handleSorting(accessor, sortOrder);
  };

  return (
    <thead className={styles.header}>
      <tr className={styles.row}>
        {columns.map(({ label, accessor, sortable, width }) => (
          <TableHeaderCell
            key={accessor as string}
            isSortable={sortable}
            isActive={accessor === sortKey}
            order={order}
            width={width}
            onClick={() => handleSortingChange(accessor)}
          >
            {label}
          </TableHeaderCell>
        ))}
      </tr>
    </thead>
  );
};
