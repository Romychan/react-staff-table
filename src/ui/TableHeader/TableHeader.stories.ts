import type { Meta, StoryObj } from '@storybook/react';

import { MOCK_COLUMNS, MOCK_SORTING } from '@tests/mock/tableMockData';

import { Users } from '@interfaces/Users';

import { TableHeader } from './TableHeader';

/** This is a component for displaying table column headers. */
const meta: Meta<typeof TableHeader<Users, keyof Users>> = {
  title: 'UI/TableHeader',
  component: TableHeader,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: { handleSorting: { action: 'handleSorting' } },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    columns: MOCK_COLUMNS,
    sorting: MOCK_SORTING,
  },
};
