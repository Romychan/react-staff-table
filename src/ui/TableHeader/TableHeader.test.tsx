import { fireEvent, screen, render } from '@testing-library/react';
import { vi } from 'vitest';

import { MOCK_COLUMNS, MOCK_SORTING } from '@tests/mock/tableMockData';

import { TableHeader } from './TableHeader';

describe('TableHeader', () => {
  it('should render correctly', () => {
    const handleClick = vi.fn();

    render(
      <TableHeader
        columns={MOCK_COLUMNS}
        sorting={MOCK_SORTING}
        handleSorting={handleClick}
      />,
    );
    const columnElement = screen.getByText(/Profile/i);

    expect(columnElement).toBeInTheDocument();
  });

  it('should handle click if sortable', () => {
    const handleClick = vi.fn();

    render(
      <TableHeader
        columns={MOCK_COLUMNS}
        sorting={MOCK_SORTING}
        handleSorting={handleClick}
      />,
    );

    const cellElement = screen.getAllByTestId('table-header-cell')[0];
    const iconElement = screen.getAllByTestId('icon')[0];

    expect(cellElement.getAttribute('class')).toMatch(/asc/gi);
    expect(iconElement.getAttribute('class')).toMatch(/icon-arrow/gi);

    fireEvent.click(cellElement);
    expect(handleClick).toHaveBeenCalledTimes(1);

    expect(cellElement.getAttribute('class')).toMatch(/desc/gi);
  });

  it('should not handle click if not sortable', () => {
    const handleClick = vi.fn();

    render(
      <TableHeader
        columns={MOCK_COLUMNS}
        sorting={MOCK_SORTING}
        handleSorting={handleClick}
      />,
    );
    const cellElement = screen.getAllByTestId('table-header-cell')[2];

    fireEvent.click(cellElement);
    expect(handleClick).toHaveBeenCalledTimes(0);
  });

  it('should match snapshot', () => {
    const handleClick = vi.fn();
    const container = render(
      <TableHeader
        columns={MOCK_COLUMNS}
        sorting={MOCK_SORTING}
        handleSorting={handleClick}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
