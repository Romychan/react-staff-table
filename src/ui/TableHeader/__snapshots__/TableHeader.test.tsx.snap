// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableHeader > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <thead
        class="_header_692233"
      >
        <tr
          class="_row_692233"
        >
          <th
            class="_cell_d2877c _sortable_d2877c _asc_d2877c"
            data-testid="table-header-cell"
          >
            <div
              class="_container_d2877c"
            >
              <svg
                class="icon icon-arrow _icon_d2877c"
                data-testid="icon"
                height="15"
                stroke="#000000"
                stroke-width="1"
                width="15"
              >
                <use
                  xlink:href="/src/assets/img/general/icons.svg#arrow"
                />
              </svg>
              ID
            </div>
          </th>
          <th
            class="_cell_d2877c _sortable_d2877c"
            data-testid="table-header-cell"
          >
            <div
              class="_container_d2877c"
            >
              <svg
                class="icon icon-arrow _icon_d2877c"
                data-testid="icon"
                height="15"
                stroke="#000000"
                stroke-width="1"
                width="15"
              >
                <use
                  xlink:href="/src/assets/img/general/icons.svg#arrow"
                />
              </svg>
              Profile
            </div>
          </th>
          <th
            class="_cell_d2877c"
            data-testid="table-header-cell"
          >
            <div
              class="_container_d2877c"
            >
              Email
            </div>
          </th>
        </tr>
      </thead>
    </div>
  </body>,
  "container": <div>
    <thead
      class="_header_692233"
    >
      <tr
        class="_row_692233"
      >
        <th
          class="_cell_d2877c _sortable_d2877c _asc_d2877c"
          data-testid="table-header-cell"
        >
          <div
            class="_container_d2877c"
          >
            <svg
              class="icon icon-arrow _icon_d2877c"
              data-testid="icon"
              height="15"
              stroke="#000000"
              stroke-width="1"
              width="15"
            >
              <use
                xlink:href="/src/assets/img/general/icons.svg#arrow"
              />
            </svg>
            ID
          </div>
        </th>
        <th
          class="_cell_d2877c _sortable_d2877c"
          data-testid="table-header-cell"
        >
          <div
            class="_container_d2877c"
          >
            <svg
              class="icon icon-arrow _icon_d2877c"
              data-testid="icon"
              height="15"
              stroke="#000000"
              stroke-width="1"
              width="15"
            >
              <use
                xlink:href="/src/assets/img/general/icons.svg#arrow"
              />
            </svg>
            Profile
          </div>
        </th>
        <th
          class="_cell_d2877c"
          data-testid="table-header-cell"
        >
          <div
            class="_container_d2877c"
          >
            Email
          </div>
        </th>
      </tr>
    </thead>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
